lib/tests/streamlit/Server_test.py

```diff
@@ -131,6 +134,58 @@ def test_healthz(self):
         self.assertEqual(503, response.code)
 
 
+class PortRotateAHundredTest(unittest.TestCase):
+    """Tests port rotates up to a hundred times then sys exits"""
+
+    def get_app(self):
+        app = mock.MagicMock()
+
+        def listen(port):
+            raise OSError(errno.EADDRINUSE, "test", "asd")
+
+        app.listen = listen
+        return app
+
+    def test_rotates_a_hundred_ports(self):
+        app = self.get_app()
+        with pytest.raises(SystemExit) as pytest_wrapped_e:
+            start_listening(app)
+            assert pytest_wrapped_e.type == SystemExit
+            assert pytest_wrapped_e.value.code == 42
+
+        assert app.listen.call_count == 100
+
+
+class PortRotateOneTest(unittest.TestCase):
+    """Tests port rotates one port"""
+
+    which_port = mock.Mock()
+
+    def get_app(self):
+        app = mock.MagicMock()
+
+        def listen(port):
+            if not port: # or port < 8502:
+                raise OSError(errno.EADDRINUSE, "test", "asd")
+            else:
+                PortRotateOneTest.which_port(port)
+
+        app.listen = listen
+        return app
+
+    @mock.patch('streamlit.config.is_manually_set')
+    @mock.patch('streamlit.config._set_option')
+    def test_rotates_one_port(self, patched__set_option, patched_is_manually_set):
+        app = self.get_app()
+
+        patched_is_manually_set.return_value = True
+
+        start_listening(app)
+
+        PortRotateOneTest.which_port.assert_called_with(8501)
+        patched__set_option.assert_called_with("server.port", 8502, "server initialization")
```

Comment:
also I want to check this function is called with this this value, and what I get logged is `not called`:
```
_mock_self = <MagicMock name='_set_option' id='4738580816'>, args = ('server.port', 8502, 'server initialization'), kwargs = {}
expected = "_set_option('server.port', 8502, 'server initialization')", actual = 'not called.'
error_message = "expected call not found.\nExpected: _set_option('server.port', 8502, 'server initialization')\nActual: not called."
```